### PR TITLE
requests patch user-agent header

### PR DESCRIPTION
**Pull request by tomashudeckiwi**
@underyx when you'll have time can we go through the file structure?

### CODE REVIEW

File: kw/platform/requests/__init__.py
Comment (underyx) [inline kw/platform/requests/__init__.py]: I'd avoid importing `*` if possible, I was thinking of this structure now:

#### `kw/platform/__init__.py`

```
from .proxies import patched_requests as requests
```

#### `kw/platform/proxies/__init__.py`

```
from .requests import patched_requests
```

#### `kw/platform/proxies/requests.py`

```
try:
    import requests
except ModuleNotFoundError:
     # something
```

…but this would be worse, cause we can't raise an import error when someone writes `from kw.platform import requests` :(

Do you see any alternatives maybe without `import *`? I don't.

File: kw/platform/requests_utils.py
Comment (underyx) [inline kw/platform/requests_utils.py]: I think the default values won't really be useful, ever, and there's some chance that people will leave these values in the User-Agent.

File: kw/platform/requests_utils.py
Comment (underyx) [inline kw/platform/requests_utils.py]: I think that with the way we're patching, we don't need to set a global for this at all. You could change this function to just return the `user_agent`, and then call it in `patch_with_user_agent()` to render the correct user-agent string specifically for the wrapper we're adding.

**Review by underyx — CHANGES_REQUESTED** [review]

File: kw/platform/requests/__init__.py
Comment (tomashudeckiwi) [reply kw/platform/requests/__init__.py]: That's what I wanted to talk about. I don't see a way around the `import *`.

Also the separate source file `requests_utils.py` seems to me as the only solution to the option of manual patching of standard `requests` library like this
```python
from kw.platform.requests_utils import patch_with_user_agent
import requests

patch_with_user_agent('some-agent')
requests.get('http://kiwi.com')
```

File: kw/platform/requests_utils.py
Comment (tomashudeckiwi) [reply kw/platform/requests_utils.py]: I agree. I just play with them during testing, I plan to leave these as empty string `""`. If they're not provided they'll be basically skipped. I don't think enforcing multiple variables would be good solution.

Comment (underyx) [reply kw/platform/requests_utils.py]: Well, I think requesting multiple variables is the only solution though. If people leave some as empty string, they'll have a broken user agent string which will fail server-side validation.

Comment (underyx) [reply kw/platform/requests_utils.py]: By the way I just realized, `APP_ENVIRONMENT` is probably a better name for this variable.

File: kw/platform/requests/__init__.py
Comment (underyx) [reply kw/platform/requests/__init__.py]: Yep, makes perfect sense, good job finding this design actually! 😊

File: kw/platform/requests_utils.py
Comment (underyx) [inline kw/platform/requests_utils.py]: Is your plan here to call each patcher one by one? Just want to clarify. So we'd later have

```
patch_with_user_agent()
patch_with_deprecation_warnings()
```

next to each other?

File: kw/platform/requests/__init__.py
Comment (underyx) [inline kw/platform/requests/__init__.py]: Oh waiiiit, I just noticed this is wrong 😧 

This patches the global requests library, which we wanted to keep intact. We should be patching specifically `kw.platform.requests`.

**Review by underyx — CHANGES_REQUESTED** [review]

File: kw/platform/requests/__init__.py
Comment (tomashudeckiwi) [reply kw/platform/requests/__init__.py]: Oh I misunderstood that completely I thought that you'll have an option to either import `kw.platform.requests` that will patch the library or original `requests` that  you'll be able to patch.

File: kw/platform/monkey.py
Comment (underyx) [inline kw/platform/monkey.py]: We might need to add a copyright notice since this seems to be almost the same as the code in ddtrace.

I asked about that here https://github.com/DataDog/dd-trace-py/pull/977#issuecomment-505832227

File: kw/platform/requests/__init__.py
Comment (underyx) [inline kw/platform/requests/__init__.py]: I'd prefer `ensure_module_is_available` or maybe even `ensure_module_exists` as they're much shorter names ☺️

File: kw/platform/requests/patch.py
Comment (underyx) [inline kw/platform/requests/patch.py]: Oh, last time when I mentioned calling it 'app environment', I meant the environment var, too.

**Review by underyx — CHANGES_REQUESTED** [review]
I'm super impressed, you brought in basically everything cool for our patching, that's awesome!

File: kw/platform/monkey.py
Comment (paveldedik) [reply kw/platform/monkey.py]: Are we going to wait for their answer before merging this?

**Comment by paveldedik** [comment]
@tomashudeckiwi I have resolved the conflicts by, well first resolving them in `pyproject.toml` and then by running `poetry update`, hope that's ok.

File: kw/platform/monkey.py
Comment (paveldedik) [inline kw/platform/monkey.py]: You cannot use the `f` strings as we want to support python 2.7 and 3.5. You need to use `format` method on strings instead. That's why tests are failing.

**Review by paveldedik — CHANGES_REQUESTED** [review]

File: test/test_requests/test_patch.py
Comment (paveldedik) [inline test/test_requests/test_patch.py]: The problem with this is that it will patch requests globally. So in all other tests, requests are patched. It will be a problem if, for example, you use `pytest.mark.parametrize` and run the test several times with different parameters - the test will pass only in the first run, in all the following runs the requests are already patched.

**Review by paveldedik — CHANGES_REQUESTED** [review]

File: kw/platform/monkey.py
Comment (underyx) [reply kw/platform/monkey.py]: No, we don't need to.

File: kw/platform/pytest/__init__.py
Comment (underyx) [inline kw/platform/pytest/__init__.py]: Could you explain what this is? 😕

File: kw/platform/monkey.py
Comment (underyx) [reply kw/platform/monkey.py]: But it seems we need to include the notice 😄so let's add that.

**Review by paveldedik — APPROVED** [review]

**Review by underyx — APPROVED** [review]

**Pull request merged**
